cub/cub/device/dispatch/dispatch_scan.cuh

```diff
@@ -189,17 +244,19 @@ struct DispatchScan
       , num_items(num_items)
       , stream(stream)
       , ptx_version(ptx_version)
+      , kernel_source(kernel_source)
+      , launcher_factory(launcher_factory)
```

Comment:
Q: Do you need a copy here? Should we `::cuda::std::move` here?

Reply:
Is it preferable then to accept the `kernel_source` parameter by rvalue reference?

My reasoning is that this would force callers to move `kernel_source` into `DispatchScan` and it wouldn't be surprising to them that the object is empty (moved from) after calling `DispatchScan`.

Reply:
> Is it preferable then to accept the kernel_source parameter by rvalue reference?

Yes. This is preferable over taking arguments as r-value reference:
```suggestion
      , kernel_source(::cuda::std::move(kernel_source))
      , launcher_factory(::cuda::std::move(launcher_factory))
```
Because then the user can choose whether to move or copy.

However, I looked up the definition of `KernelSource` and `KernelLauncherFactory` and they contain no data, so copying is fine here.

Reply:
Understood. I am unresolving this conversation just to ask a C++/API design question:

Let us assume the following:

1. `kernel_source` had a move constructor
2. We made the change you suggested above

Now, if someone created a `DispatchScan` as follows, isn't there a risk that they see an unexpected side-effect?

```cpp
auto kernel_source = make_kernel_source(...);
auto scan = DispatchScan(..., kernel_source, ...);
// now kernel_source has been moved from even though I don't necessarily expect that?
```

Reply:
If the type of `kernel_source` had only a move constructor (and no copy constructor), your code would not compile. You would need to explicitly call `move(kernel_source)` in your snippet.

If the type of `kernel_source` had a move and a copy constructor, calling `Dispatch(...)` in your snippet would pull a copy of `kernel_source`, and the copy would later be moved into the data member.

Reply:
Got it - thank you for the clarification!